### DISCUSSION TITLE
Application Health Metrics

### DISCUSSION THREAD

**Discussion by imagiczhang** [MobileNativeFoundation/discussions#125] [General]
I'm curious in how other engineers collect, proceed and monitor the signal of application health including but not limited to application's performance, reliability, battery efficiency and etc. What are useful metric which can accurately reflect health problems? How to reduce the noises when collection metrics data?

**Comment by PiotrLopata** [comment]
Hello Mobile folks!

I see this question stays unanswered for a quite a bit and still nobody has introduced an efficient way to monitor our apps health. I'm going to tell a story about our (Allegro) try to handle this subject. We didn't decide to make our solution official inside our iOS community, cause it was PoC (Proof of Concept), but I believe it's worth sharing and may be inspiring for some of You.

## Where are we?

We're using third party library, which measures our performance indicators. We're checking HTTP errors, network errors, response times etc. Those measures are connected with our Allegro SLA (Service Level Agreement). But this library is quite expensive, so we came up with an idea to create something similar with use of our resources.

## Solution proposal 

* Use of analytics events sent from our app
* Manage the data inside new service
    * Create proper metric name from event values
    * Send measured data into metric collector on the server side - every minute

## Which events could we use

* First thought - we already collect some events. All used by our analysts team. For example successful/failed login attempts
* New mechanism inside our networking framework, which allows us to monitor requests made by the app
    * Remotely configured network extension, which creates new analitycs event with specified data. We're using pattern from backend services monitoring with data such as endpoint path, status code with its family

## Metric we use

* How many events were processed per second each minute

## Quality of our solution

1. When we wanted to create some metric with already in use analytics event, we faced very time consuming problem. For example measuring login conversion. Math looks simple -> successful logins divided by started authentications. We observed, we've got more started login tries then the results, so we discovered quickly, there is an event missing for login try by using keyboard `Return` button tap. Adding such an event costs are:
    * development
    * tests
    * release + rollout
    
    So as You see, all cases should be handled to have proper results, if not, it would take time.
2. Internet connection problems - You never know when gathered event would be send into our service
3. Our implementation - we are sending batched events. When required amount of events is gathered, they're sent into our service. Events that don't fit our one minute window cause collection noises
4. In order to make this metrics reliable, You should have a significant traffic inside the app. It helps to reduce collection noises and avoid cases like: single user with failed authentication causes alarm

## Why We don't use this approach?

We're using third party library, which allows us to check some data (we didn't defined them first) and we tried to replace it. When this PoC was made, we were comparing current library with created one. It was a mistake. We should take some thoughts about what data is critical for us, and then think how could we gather them.

## What's next?

We're going to take another try, to find or make most suitable solution for us, but this time with defining our needs in a first place.
